File: pramukhraj-foods-server/pramukhraj/Entities/Product/Product.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace pramukhraj.Entities.Product
{
        [Table("Products")]
        [Index(nameof(Slug), IsUnique = true)]
        [Index(nameof(CategoryId))]
        public class Product
        {
            [Key]
            [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
            public Guid Id { get; set; }

            [Required]
            public Guid CategoryId { get; set; }

            [ForeignKey(nameof(CategoryId))]
            public ProductCategory Category { get; set; } = default!;

            [Required]
            [MaxLength(255)]
            public string Name { get; set; } = string.Empty;

            [Required]
            [MaxLength(255)]
            public string Slug { get; set; } = string.Empty;

            [MaxLength(500)]
            public string ShortDescription { get; set; } = string.Empty;

            // No MaxLength defaults to Postgres 'text' type, good for long content
            public string Description { get; set; } = string.Empty;

            [MaxLength(100)]
            public string Brand { get; set; } = "Pramukhraj";

            public bool IsFeatured { get; set; }
            public bool IsBestSeller { get; set; }
            public bool IsTrending { get; set; }
            public bool IsNewArrival { get; set; }
            public bool IsActive { get; set; } = true;

            [MaxLength(100)]
            public string CountryOfOrigin { get; set; } = "India";

            public bool IsVegetarian { get; set; } = true;

            [MaxLength(100)]
            public string ShelfLife { get; set; } = string.Empty;

            [MaxLength(500)]
            public string StorageInstruction { get; set; } = string.Empty;

            public string Ingredients { get; set; } = string.Empty;

            public string NutritionalInformation { get; set; } = string.Empty;

            [MaxLength(100)]
            public string? Barcode { get; set; }

            [MaxLength(255)]
            public string? MetaTitle { get; set; }

            [MaxLength(500)]
            public string? MetaDescription { get; set; }

            [MaxLength(500)]
            public string? MetaKeywords { get; set; }

            public DateTime CreatedOn { get; set; } = DateTime.UtcNow;

            public DateTime UpdatedOn { get; set; } = DateTime.UtcNow;

            public ICollection<ProductImage> Images { get; set; } = new List<ProductImage>();
            public ICollection<ProductVariant> Variants { get; set; } = new List<ProductVariant>();
            public ICollection<ProductTag> Tags { get; set; } = new List<ProductTag>();
        }
}

// Disconted price (MRP-Price)/MRP * 100

File: pramukhraj-foods/src/model/Product.ts
// ─── Product Image ────────────────────────────────────────────────────────────

export class ProductImage {
  id: string = ''
  productId: string = ''
  imageUrl: string = ''
  altText: string | null = null
  isPrimary: boolean = false
  displayOrder: number = 0
}

// ─── Product Variant ──────────────────────────────────────────────────────────

export class ProductVariant {
  id: string = ''
  productId: string = ''
  name: string = ''
  sku: string = ''
  price: number = 0       // *required
  mrp: number = 0         // *required
  stockQuantity: number = 0 // *required
  weight: number = 0
  weightUnit: string = 'gm'
  isDefault: boolean = false
  isActive: boolean = true
}

// ─── Product Tag ──────────────────────────────────────────────────────────────

export class ProductTag {
  id: string = ''
  productId: string = ''
  name: string = ''
}

// ─── Product ──────────────────────────────────────────────────────────────────

export class Product {
  id: string = ''
  categoryId: string = ''           // select field  *required
  name: string = ''                 // *required
  shortDescription: string = ''
  description: string = ''
  brand: string = 'Pramukhraj'      // *required
  isFeatured: boolean = false
  isBestSeller: boolean = false
  isTrending: boolean = false
  isNewArrival: boolean = false
  isActive: boolean = true
  weight: number = 0                // *required (> 0)
  weightUnit: string = 'gm'
  countryOfOrigin: string = 'India' // *required
  isVegetarian: boolean = true
  shelfLife: string = ''            // empty = lifetime good
  storageInstruction: string = ''
  ingredients: string = ''
  nutritionalInformation: string = ''
  barcode: string | null = null
  createdOn: string = ''
  updatedOn: string = ''
  images: ProductImage[] = []       // *required, only one isPrimary
  variants: ProductVariant[] = []
  tags: ProductTag[] = []

  static calculatedDiscountPercentage(mrp: number, price: number): number {
    if (mrp <= 0 || price >= mrp) return 0
    return Number((((mrp - price) / mrp) * 100).toFixed(2))
  }
}

// ─── Weight unit options ──────────────────────────────────────────────────────

export const WEIGHT_UNITS = ['gm', 'kg', 'ml', 'L', 'piece'] as const
export type WeightUnit = (typeof WEIGHT_UNITS)[number]

// ─── Step definitions ─────────────────────────────────────────────────────────

export interface ProductFormStep {
  id: number
  label: string
  description: string
}

export const PRODUCT_FORM_STEPS: ProductFormStep[] = [
  { id: 1, label: 'Basic Info', description: 'Name, category, brand & core flags' },
  { id: 2, label: 'Details', description: 'Weight, origin, ingredients & nutrition' },
  { id: 3, label: 'Variants', description: 'Pricing, SKU & stock per variant' },
  { id: 4, label: 'Tags', description: 'Searchable labels' },
  { id: 5, label: 'Images', description: 'Gallery & primary image' },
  { id: 6, label: 'Review', description: 'Confirm everything before saving' },
]